pkg/cluster/resources.go

```diff
@@ -126,44 +126,82 @@ func (c *Cluster) createConnectionPooler(lookup InstallFunction) (*ConnectionPoo
-	if err != nil {
-		msg = "could not generate deployment for connection pooler: %v"
-		return nil, fmt.Errorf(msg, err)
-	}
+		deploymentSpec, err := c.generateConnectionPoolerDeployment(&c.Spec, Master)
+		if err != nil {
+			msg = "could not generate deployment for connection pooler: %v"
+			return nil, fmt.Errorf(msg, err)
+		}
 
-	// client-go does retry 10 times (with NoBackoff by default) when the API
-	// believe a request can be retried and returns Retry-After header. This
-	// should be good enough to not think about it here.
-	deployment, err := c.KubeClient.
-		Deployments(deploymentSpec.Namespace).
-		Create(context.TODO(), deploymentSpec, metav1.CreateOptions{})
+		// client-go does retry 10 times (with NoBackoff by default) when the API
+		// believe a request can be retried and returns Retry-After header. This
+		// should be good enough to not think about it here.
+		deployment, err := c.KubeClient.
+			Deployments(deploymentSpec.Namespace).
+			Create(context.TODO(), deploymentSpec, metav1.CreateOptions{})
 
-	if err != nil {
-		return nil, err
-	}
+		if err != nil {
+			return nil, err
+		}
 
-	serviceSpec := c.generateConnectionPoolerService(&c.Spec)
-	service, err := c.KubeClient.
-		Services(serviceSpec.Namespace).
-		Create(context.TODO(), serviceSpec, metav1.CreateOptions{})
+		serviceSpec := c.generateConnectionPoolerService(&c.Spec, Master)
+		service, err := c.KubeClient.
+			Services(serviceSpec.Namespace).
+			Create(context.TODO(), serviceSpec, metav1.CreateOptions{})
+
+		if err != nil {
+			return nil, err
+		}
+		c.ConnectionPooler = &ConnectionPoolerObjects{
+			Deployment: deployment,
+			Service:    service,
+		}
+		c.logger.Debugf("created new connection pooler %q, uid: %q",
+			util.NameFromMeta(deployment.ObjectMeta), deployment.UID)
 
-	if err != nil {
-		return nil, err
 	}
 
-	c.ConnectionPooler = &ConnectionPoolerObjects{
-		Deployment: deployment,
-		Service:    service,
+	if c.needReplicaConnectionPooler() {
+
+		repldeploymentSpec, err := c.generateConnectionPoolerDeployment(&c.Spec, Replica)
+		if err != nil {
+			msg = "could not generate deployment for connection pooler: %v"
+			return nil, fmt.Errorf(msg, err)
+		}
+
+		// client-go does retry 10 times (with NoBackoff by default) when the API
+		// believe a request can be retried and returns Retry-After header. This
+		// should be good enough to not think about it here.
+		repldeployment, err := c.KubeClient.
+			Deployments(repldeploymentSpec.Namespace).
+			Create(context.TODO(), repldeploymentSpec, metav1.CreateOptions{})
+
+		if err != nil {
+			return nil, err
+		}
+
+		replServiceSpec := c.generateConnectionPoolerService(&c.Spec, Replica)
+		replService, err := c.KubeClient.
+			Services(replServiceSpec.Namespace).
+			Create(context.TODO(), replServiceSpec, metav1.CreateOptions{})
+
+		if err != nil {
+			return nil, err
+		}
+		c.ConnectionPooler = &ConnectionPoolerObjects{
```

Comment:
does this work, if e.g. master connection pooler is enabled? does this not overwrite?